src/Microsoft.ML.Dnn/DnnRetrainTransform.cs

```diff
@@ -811,18 +812,22 @@ private sealed class Mapper : MapperBase
                             throw Contracts.Except($"Input shape mismatch: Input '{_parent._inputs[i]}' has shape {originalShape.ToString()}, but input data is of length {typeValueCount}.");
 
                         // Fill in the unknown dimensions.
-                        var l = new int[originalShape.NDim];
-                        for (int ishape = 0; ishape < originalShape.NDim; ishape++)
-                            l[ishape] = originalShape[ishape] == -1 ? (int)d : originalShape[ishape];
+                        var l = new int[originalShape.ndim];
+                        for (int ishape = 0; ishape < originalShape.ndim; ishape++)
+                        {
+                            var dims = originalShape.dims;
```

Comment:
I think this line can be lifted outside of the `for` loop. No reason to do it every time through the loop.